// sdapi/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tense = exports.Person = exports.Mood = exports.Language = exports.Gender = exports.Form = exports.CNumber = void 0;
var Gender;
(function (Gender) {
    Gender["Masculine"] = "m";
    Gender["Femeline"] = "f";
    Gender["Neutral"] = "n";
})(Gender || (Gender = {}));
exports.Gender = Gender;
var Language;
(function (Language) {
    Language["English"] = "en";
    Language["Spanish"] = "es";
})(Language || (Language = {}));
exports.Language = Language;
var Person;
(function (Person) {
    Person["First"] = "1st";
    Person["Second"] = "2nd";
    Person["Third"] = "3rd";
})(Person || (Person = {}));
exports.Person = Person;
var CNumber;
(function (CNumber) {
    CNumber["Singular"] = "sig";
    CNumber["Plural"] = "plr";
})(CNumber || (CNumber = {}));
exports.CNumber = CNumber;
var Tense;
(function (Tense) {
    Tense["Present"] = "present";
    Tense["Preterite"] = "preterite";
    Tense["Imperfect"] = "imperfect";
    Tense["Conditional"] = "conditional";
    Tense["Imperfect2"] = "imperfect2";
    Tense["Future"] = "future";
    Tense["Affirmative"] = "affirmative";
    Tense["Negative"] = "negative";
    Tense["Past"] = "past";
})(Tense || (Tense = {}));
exports.Tense = Tense;
var Mood;
(function (Mood) {
    Mood["Indicative"] = "ind";
    Mood["Subjunctive"] = "sub";
    Mood["Imperative"] = "imp";
})(Mood || (Mood = {}));
exports.Mood = Mood;
// Not a linguist, just trying to separate them in the program
var Form;
(function (Form) {
    Form["Simple"] = "simp";
    Form["Progressive"] = "prog";
    Form["Perfect"] = "perf";
})(Form || (Form = {}));
exports.Form = Form;
